plugin/cql/src/main/java/org/opencds/cqf/ruler/cql/KnowledgeArtifactProcessor.java

```diff
@@ -171,15 +194,100 @@ public MetadataResource approve(IdType idType, Date approvalDate, String artifac
+		}
+		// Ensure only one resource exists with this URL
+		Bundle existingArtifactsForUrl = searchResourceByUrl(draftVersionUrl, fhirDal);
+		if(existingArtifactsForUrl.getEntry().size() != 0){
+			throw new UnprocessableEntityException(
+				String.format("A draft of Program '%s' already exists with version: '%s'. Only one draft of a program version can exist at a time.", baseArtifact.getUrl(), draftVersionUrl));
+		}
+		List<MetadataResource> resourcesToCreate = createDraftsOfArtifactAndRelated(baseArtifact, fhirDal, version, new ArrayList<MetadataResource>());
+		Bundle transactionBundle = new Bundle()
+			.setType(Bundle.BundleType.TRANSACTION);
+		List<IdType> urnList = resourcesToCreate.stream().map(res -> new IdType("urn:uuid:" + UUID.randomUUID().toString())).collect(Collectors.toList());
+		for(int i = 0; i < resourcesToCreate.size(); i++){
+			KnowledgeArtifactAdapter<MetadataResource> newResourceAdapter = new KnowledgeArtifactAdapter<MetadataResource>(resourcesToCreate.get(i));
+			updateUsageContextReferencesWithUrns(resourcesToCreate.get(i), resourcesToCreate, urnList);
+			updateRelatedArtifactUrlsWithNewVersions(newResourceAdapter, draftVersion);
+			MetadataResource updateIdForBundle = newResourceAdapter.copy();
+			updateIdForBundle.setId(urnList.get(i));
+			transactionBundle.addEntry(createEntry(updateIdForBundle));
+		}
+		return transactionBundle;
+	}
+	private void updateUsageContextReferencesWithUrns(MetadataResource newResource, List<MetadataResource> resourceListWithOriginalIds, List<IdType> idListForTransactionBundle){
+		List<UsageContext> useContexts = newResource.getUseContext();
+		for(UsageContext useContext : useContexts){
+			if(useContext.hasValueReference()){
+				Reference useContextRef = useContext.getValueReference();
+				if(useContextRef != null){
+					resourceListWithOriginalIds.stream()
+						.filter(resource -> (resource.getClass().getSimpleName() + "/" + resource.getIdElement().getIdPart()).equals(useContextRef.getReference()))
+						.findAny()
+						.ifPresent(resource -> {
+							int indexOfDraftInIdList = resourceListWithOriginalIds.indexOf(resource);
+							useContext.setValue(new Reference(idListForTransactionBundle.get(indexOfDraftInIdList)));
+						});
+				}
+			}
 		}
-
-		if (version.contains(".")) {
+	}
+	private void updateRelatedArtifactUrlsWithNewVersions(KnowledgeArtifactAdapter<MetadataResource> newResourceAdapter, String draftVersion){
+			// For each Resource relatedArtifact, update the version of the reference.
+			newResourceAdapter.getRelatedArtifact().stream()
+				.filter(ra -> ra.hasResource()).collect(Collectors.toList())
+				.replaceAll(ra -> ra.setResource(Canonicals.getUrl(ra.getResource()) + "|" + draftVersion));
+	}
+	private void checkIfVersionIsValid(String version) throws UnprocessableEntityException{
+		if (version == null || version.isEmpty()) {
+			throw new UnprocessableEntityException("The version argument is required");
```

Comment:
@sliver007 I changed the exception type here so that it matched the rest of the version validity errors (and also threw something extending `FHIRException`), but I'm happy to make them all `InvalidOperatorArgument` if that's preferred

Reply:
Much better! Thanks.